### ISSUE TITLE
XML for `ietf-alarms` notification is invalid

### ISSUE THREAD

**Issue by unmanned-player** [sysrepo/sysrepo#1069]
(put "x" in "[]" if you already tried following)
* [x] Did you check if this is a duplicate issue?
* [x] Did you test it on the latest sysrepo devel branch?

### Description
With the latest version of libyang, libnetconf2, Netopeer2 and sysrepo we're unable to send ietf-alarm notifications. The alarm notification file it generates has an error and a call to `sr_event_notif_send` returns internal error.

This is what we're trying to send:

```
send alarm
  x = /ietf-alarms:alarm-notification/resource, v = pingpong
  x = /ietf-alarms:alarm-notification/alarm-type-id, v = my-alarms:pingpong
  x = /ietf-alarms:alarm-notification/alarm-type-qualifier, v = pingpong
  x = /ietf-alarms:alarm-notification/alt-resource, v = pingpong
  x = /ietf-alarms:alarm-notification/impacted-resource, v = pingpong
  x = /ietf-alarms:alarm-notification/root-cause-resource, v = pingpong
  x = /ietf-alarms:alarm-notification/time, v = 2018-03-05T03:05:00+00:00
  x = /ietf-alarms:alarm-notification/perceived-severity, v = critical
  x = /ietf-alarms:alarm-notification/alarm-text, v = blah blah
also_send_ietf_alarm: Sysrepo-internal error
```

I've attached the yangs and the alarm.c source code in the zip for reference. [yang.zip](https://github.com/sysrepo/sysrepo/files/1779414/yang.zip)

### Steps to Reproduce
1. Compile `alarm.c` (preferably as another example in sysrepo).
2. Make sure to install `ietf-alarms@2018-02-01.yang` and 
2. Start `sysrepod` and `sysrepo-plugind`.
3. Start `netopeer-cli`, connect to the local-host, and `subscribe`.
4. Run `alarm` on the same system where others are running.

**Expected behavior:**
We expect to see that that alarm notification is received by the CLI.

**Actual behavior:**
The `sysrepod` reports an error parsing the generated notification and doesn't send any notification to the CLI client.

```
[DBG] (rp_event_notif_req_process:3027) Processing event notification request (/ietf-alarms:alarm-notification).
[DBG] (ac_set_identity:155) Switching identity to UID='0' and GID='0' (username: root).
[DBG] (dm_load_data_tree_file:1105) Loaded module ietf-alarms: mtime sec=1520220113 nsec=341106401
[DBG] (dm_load_data_tree_file:1189) Usage count ietf-alarms incremented (value=1)
[INF] (dm_load_data_tree_file:1195) Data file /etc/sysrepo/data/ietf-alarms.running loaded successfully
[DBG] (dm_get_data_info_internal:2736) Module ietf-alarms has been loaded
[DBG] (dm_validate_procedure_content:6072) Validation succeeded for 'ietf-alarms' module
[DBG] (dm_get_data_info_internal:2689) Module ietf-alarms already loaded
[DBG] (np_store_event_notification:1673) Storing notification '/ietf-alarms:alarm-notification' generated on '1520220183'.
[DBG] (ac_set_identity:155) Switching identity to UID='0' and GID='0' (username: root).
[DBG] (sr_locking_set_lock_fd:789) File /etc/sysrepo/data/notifications//ietf-alarms/2018-03-05_14-20.xml has been locked
[ERR] (dm_ly_log_cb:1319) libyang error: Invalid NameStartChar of the attribute.
[ERR] (np_load_data_tree:408) Parsing data from file '/etc/sysrepo/data/notifications//ietf-alarms/2018-03-05_14-20.xml' failed: Invalid NameStartChar of the attribute.
[DBG] (sr_locking_set_unlock_close_fd:858) File /etc/sysrepo/data/notifications//ietf-alarms/2018-03-05_14-20.xml (fd = 11) has been unlocked
[ERR] (np_store_event_notification:1696) Unable to load notification store data for module 'ietf-alarms'.
[ERR] (rp_event_notif_req_process:3085) Failed to save event notification
```

Note, the generated XML (/etc/sysrepo/data/notifications//ietf-alarms/2018-03-05_14-20.xml) is broken near the `resource` tag. See:

```
<notifications xmlns="urn:ietf:params:xml:ns:yang:sysrepo-notification-store">
  <notification>
    <xpath>/ietf-alarms:alarm-notification</xpath>
    <generated-time>2018-03-05T14:22:51+11:00</generated-time>
    <logged-time>1698161834</logged-time>
    <data>
      <alarm-notification xmlns="urn:ietf:params:xml:ns:yang:ietf-alarms">
        <resource"(!error!)"
```

Thanks
Anandu

**Comment by michalvasko** [comment]
Hi Anandu,
it should be fixed with the pull request and most recent *libyang*. However, you need to delete the notifications file you mentioned because as you see, it is invalid. The newly created files should be correct.

Regards,
Michal

**Comment by unmanned-player** [comment]
I tried it with the latest patch but the XML seems a bit odd.

```
<notifications xmlns="urn:ietf:params:xml:ns:yang:sysrepo-notification-store">
  <notification>
    <xpath>/ietf-alarms:alarm-notification</xpath>
    <generated-time>2018-03-09T12:22:30+01:00</generated-time>
    <logged-time>1731999757</logged-time>
    <data>&lt;alarm-notification xmlns="urn:ietf:params:xml:ns:yang:ietf-alarms"&gt;
  &lt;resource&gt;pingpong&lt;/resource&gt;
  &lt;alarm-type-id xmlns:example-a="http://www.example.com/alarms"&gt;example-a:pingpong&lt;/alarm-type-id&gt;
  &lt;alarm-type-qualifier&gt;pingpong&lt;/alarm-type-qualifier&gt;
  &lt;alt-resource&gt;pingpong&lt;/alt-resource&gt;
  &lt;impacted-resource&gt;pingpong&lt;/impacted-resource&gt;
  &lt;root-cause-resource&gt;pingpong&lt;/root-cause-resource&gt;
  &lt;time&gt;2018-03-09T12:22:30+01:00&lt;/time&gt;
  &lt;perceived-severity&gt;critical&lt;/perceived-severity&gt;
  &lt;alarm-text&gt;blah blah&lt;/alarm-text&gt;
&lt;/alarm-notification&gt;
</data>
  </notification>
</notifications>
```

I was expecting a proper XML instead.

**Comment by michalvasko** [comment]
Hi,
yeah, this is an expected side-effect of the changes I made. Is this a problem? It should make no difference to *sysrepo*. I will try to think about a way to keep the notification itself XML-formatted but I don't really find this a problem.

Regards,
Michal

**Comment by michalvasko** [comment]
Hi,
try it now, it should be just like before. However, the XML files are not really supposed to be human-readable.

Regards,
Michal

**Comment by unmanned-player** [comment]
Hello Michal,

Sorry for the delayed reply. I was a bit stuck with other activities.

The patch isn't still sending alarms properly. It does generate the XML and leave it in `/etc/sysrepo/data/notifications/ietf-alarms`. But when netopeer2 server tries to send the notification it fails when trying to parse the XML back.

```
netopeer2-server[3975]: Session 1 received a realtime notification "/ietf-alarms:alarm-notification" (1522039897).
netopeer2-server[3975]: unable to write notification
```
For now, I've patched it in libyang. I changed printer_xml.c and parser.c to ignore LY_TYPE_UNION on special case. I've attached the patch [1069.patch.zip](https://github.com/sysrepo/sysrepo/files/1849554/1069.patch.zip) for reference. I use the same files to test though.

Thanks,
Anandu
